internal/garbagecollection/garbage_collector.go

```diff
@@ -0,0 +1,94 @@
+package garbagecollection
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/metadata"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/operator-framework/catalogd/api/core/v1alpha1"
+)
+
+var _ manager.Runnable = (*GarbageCollector)(nil)
+
+// GarbageCollector is an implementation of the manager.Runnable
+// interface for running garbage collection on the Catalog content
+// cache that is served by the catalogd HTTP server. It runs in a loop
+// and will ensure that no cache entries exist for Catalog resources
+// that no longer exist. This should only clean up cache entries that
+// were missed by the handling of a DELETE event on a Catalog resource.
+type GarbageCollector struct {
+	CachePath      string
+	Logger         logr.Logger
+	MetadataClient metadata.Interface
+	Interval       time.Duration
+}
+
+// Start will start the garbage collector. It will always run once on startup
+// and loop until context is canceled after an initial garbage collection run.
+// Garbage collection will run again every X amount of time, where X is the
+// supplied garbage collection interval.
+func (gc *GarbageCollector) Start(ctx context.Context) error {
+	// Run once on startup
+	removed, err := runGarbageCollection(ctx, gc.CachePath, gc.MetadataClient)
+	if err != nil {
+		gc.Logger.Error(err, "running garbage collection")
+	}
+	if len(removed) > 0 {
+		gc.Logger.Info("removed stale cache entries", "removed entries", removed)
+	}
+
+	// Loop until context is canceled, running garbage collection
+	// at the configured interval
+	for {
```

Comment:
FYI take a peek at the `wait` package from k8s for prior art and some of the edge cases

Reply:
Can do in a future follow up. Thanks!